File: presets/sofast-hosts/src/no-view/hosts-read.ts
/// <reference types="node" />
import { ctx, done, log, onError } from '@shared/lib/host-node'
import type { HostsReadArgs, HostsReadResult } from '../core/script-types'
import { listBackups, readBackup, readHostsFile, resolveHostsPath } from './_hosts-file'

/**
 * script 命令：读取系统 hosts 文件。
 *
 * package.json 里声明为 { "name": "hosts-read", "mode": "script" }，
 * View 侧用 Backend.run('hosts-read', args) 调用；产物必须是 dist/hosts-read.mjs。
 *
 * args:
 *   { backups: true }              → 顺带返回磁盘备份列表
 *   { backup: 'hosts-2026-…txt' }  → 读取指定备份的内容（不读系统文件）
 */

function fail(message: string): HostsReadResult {
  return {
    ok: false,
    path: resolveHostsPath(),
    content: '',
    size: 0,
    mtime: 0,
    bom: false,
    encoding: 'utf8',
    writable: false,
    platform: process.platform,
    error: message,
  }
}

onError()

void (async () => {
  const { args, dataPath } = ctx()
  const opts = (args ?? {}) as HostsReadArgs
  // 备份目录固定为 <数据目录>/backups（N2：安装目录只读）。
  // 如快下 dataPath 兜底成 <插件目录>/data，位置与历史版本完全一致。
  const baseDir = dataPath

  try {
    if (opts.backup) {
      const content = readBackup(baseDir, String(opts.backup))
      const info = listBackups(baseDir).find((b) => b.name === opts.backup)
      log('hosts-read: 读取备份', { name: opts.backup, size: content.length })
      done({
        ok: true,
        path: resolveHostsPath(),
        content,
        size: info?.size ?? content.length,
        mtime: info?.mtime ?? 0,
        bom: false,
        encoding: 'utf8',
        writable: false,
        platform: process.platform,
      } satisfies HostsReadResult)
      return
    }

    const result = readHostsFile()
    if (opts.backups) result.backups = listBackups(baseDir)
    log('hosts-read: 完成', { path: result.path, size: result.size, writable: result.writable, encoding: result.encoding })
    done(result)
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err)
    log('hosts-read: 失败', { message }, 'error')
    done(fail(message))
  }
})()

File: plugins/host-manager/src/bin/hosts_read.rs
//! 命令 `hosts-read`（script）：读取系统 hosts 文件（v1 `src/no-view/hosts-read.ts`）。
//!
//! args:
//!   `{ backups: true }`              → 顺带返回磁盘备份列表
//!   `{ backup: 'hosts-2026-…txt' }`  → 读取指定备份的内容（不读系统文件）

use launcher_plugin_host_manager::{list_backups, platform_string, read_backup, read_hosts_file, resolve_hosts_path};
use launcher_plugin_sdk::{json, Context, Level, Result, Value};

fn main() {
    launcher_plugin_sdk::run(dispatch);
}

fn dispatch(ctx: &Context) -> Result<()> {
    // 备份目录固定为 <数据目录>/backups（N2：安装目录只读）
    let data_path = ctx.data_path().to_path_buf();
    let args = ctx.raw_args().clone();

    if let Some(name) = args.get("backup").and_then(Value::as_str) {
        return match read_backup(&data_path, name) {
            Ok(content) => {
                let info = list_backups(&data_path).into_iter().find(|entry| entry.name == name);
                ctx.log(
                    &format!("hosts-read: 读取备份 {name}（{} 字节）", content.len()),
                    None,
                    Level::Info,
                )?;
                ctx.done(json!({
                    "ok": true,
                    "path": resolve_hosts_path().to_string_lossy(),
                    "content": content,
                    "size": info.as_ref().map(|entry| entry.size).unwrap_or(content.len() as u64),
                    "mtime": info.map(|entry| entry.mtime).unwrap_or(0),
                    "bom": false,
                    "encoding": "utf8",
                    "writable": false,
                    "platform": platform_string(),
                }))
            }
            Err(message) => {
                ctx.log(&format!("hosts-read: 失败 {message}"), None, Level::Error)?;
                ctx.done(fail_result(&message))
            }
        };
    }

    let mut result = read_hosts_file(None);
    if args.get("backups").and_then(Value::as_bool) == Some(true) {
        result.backups = Some(list_backups(&data_path));
    }
    ctx.log(
        &format!(
            "hosts-read: 完成（path={} size={} writable={} encoding={}）",
            result.path, result.size, result.writable, result.encoding
        ),
        None,
        Level::Info,
    )?;
    ctx.done(serde_json::to_value(&result).unwrap_or(Value::Null))
}

/// 与 v1 的 `fail()` 同形状：读失败也回一份完整的 `HostsReadResult`（调用方只判 `ok`）。
fn fail_result(message: &str) -> Value {
    json!({
        "ok": false,
        "path": resolve_hosts_path().to_string_lossy(),
        "content": "",
        "size": 0,
        "mtime": 0,
        "bom": false,
        "encoding": "utf8",
        "writable": false,
        "platform": platform_string(),
        "error": message,
    })
}
